### PR TITLE
Fixes 1501: add API to list snapshots

### PR DESCRIPTION
**Pull request by rverdile**
## Summary
Adds an API endpoint to list the snapshots of a repository: `repositories/:uuid/snapshots/`
## Testing steps
1. Using the API, create a repository, and in the body include the field `snapshot: true`
2. GET request `repositories/{new repository UUID}/snapshots/`
3. Should see the snapshot returned, including pagination metadata.

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-1501

File: pkg/dao/snapshots.go [hunk 1]
```diff
@@ -18,16 +19,37 @@ func (sDao snapshotDaoImpl) Create(s *Snapshot) error {
...
+		snapshotModelToApi(snapshots[i], &repos[i])
+	}
+	return repos
+}
+
+func snapshotModelToApi(model Snapshot, resp *api.SnapshotResponse) {
+	resp.CreatedAt = model.CreatedAt
+	resp.DistributionPath = model.DistributionPath
+	resp.PackageCount = int(model.ContentCounts["rpm.package"])
+	resp.ErrataCount = int(model.ContentCounts["rpm.advisory"])
```
Comment (jlsherrill) [inline pkg/dao/snapshots.go:54]: i had kinda imagined we'd return ContentCounts as is (essentially a map string -> int).   This makes adding types in the future simpler.  Any Thoughts?

Comment (rverdile) [reply pkg/dao/snapshots.go:54]: Meaning the response looks like this?
```
{
   "created_at": [timestamp],
   "distribution_path": "aaa/bbb",
   "content_counts": {
        "rpm.package": 30,
        "rpm.advisory": 4,
        "rpm.packagegroup": 2,
         ...,
   }
}
```

File: pkg/handler/repositories.go [hunk 2]
```diff
@@ -398,6 +399,30 @@ func (rh *RepositoryHandler) introspect(c echo.Context) error {
...
+func (rh *RepositoryHandler) listSnapshots(c echo.Context) error {
+	uuid := c.Param("uuid")
+	pageData := ParsePagination(c)
+	filterData := ParseFilters(c)
+	snapshots, totalSnaps, err := rh.DaoRegistry.Snapshot.List(uuid, pageData, filterData)
+	if err != nil {
+		return ce.NewErrorResponse(ce.HttpCodeForDaoError(err), "Error listing repositories", err.Error())
+	}
+	return c.JSON(200, setCollectionResponseMetadata(&snapshots, c, totalSnaps))
+}
```
[the remark this answers is not in the document]
Comment (rverdile) [reply pkg/handler/repositories.go:424]: sure, so a snapshot handler, but the route is still under `/repositories/`?

File: pkg/dao/snapshots.go (hunk 1 above)
Comment (jlsherrill) [reply pkg/dao/snapshots.go:54]: correct.  Its passing the key straight from pulp, which is kinda ugly, but its tucked into a nice little structure.  

If we wanted a cleaner key, we could translate from  'rpm.package' to 'package' or something similar.   I try to think about the future if we have some other type of repository that isn't yum

File: pkg/handler/repositories.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlsherrill) [reply pkg/handler/repositories.go:424]: correct, basically everything under  `/repositories/uuid/snapshots/`

File: pkg/dao/snapshots.go (hunk 1 above)
Comment (rverdile) [reply pkg/dao/snapshots.go:54]: Got it. I'm okay with returning it that way.

**Review by jlsherrill — APPROVED** [review]

**Comment by swadeley** [comment]
Hi

I created a repo with snapshot=true

I can see the tasks (one is for introspection, one for snapshot):

```
In [3]: app.content_sources.rest_client.tasks_api.list_tasks()
2023-06-23 15:05:48.753 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=o4K7ve1lVtam5a28d7wagxZQ8sH5yOSk, params=[]
Out[3]: 
{'data': [{'created_at': '2023-06-23T14:05:44Z',
           'ended_at': '2023-06-23T14:05:44Z',
           'error': '',
           'org_id': '3340851',
           'status': 'completed',
           'uuid': 'b8b72afd-0859-491b-9182-4de101fe0dab'},
          {'created_at': '2023-06-23T14:05:44Z',
           'ended_at': '',
           'error': '',
           'org_id': '3340851',
           'status': 'running',
           'uuid': 'e6fb679a-7eac-46b7-b2e6-78a077005c5e'}],
 'links': {'first': '/api/content-sources/v1/tasks/?limit=100&offset=0',
           'last': '/api/content-sources/v1/tasks/?limit=100&offset=0'},
 'meta': {'count': 2, 'limit': 100, 'offset': 0}}
```

A bit later (see second task's "ended_at" and "status" values):
```
In [5]: app.content_sources.rest_client.tasks_api.list_tasks()
2023-06-23 15:06:47.160 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=Ijz8qGoQSYFmBNscX2zLXgQ8kVb6RSc7, params=[]
Out[5]: 
{'data': [{'created_at': '2023-06-23T14:05:44Z',
           'ended_at': '2023-06-23T14:05:44Z',
           'error': '',
           'org_id': '3340851',
           'status': 'completed',
           'uuid': 'b8b72afd-0859-491b-9182-4de101fe0dab'},
          {'created_at': '2023-06-23T14:05:44Z',
           'ended_at': '2023-06-23T14:06:19Z',
           'error': '',
           'org_id': '3340851',
           'status': 'completed',
           'uuid': 'e6fb679a-7eac-46b7-b2e6-78a077005c5e'}],
 'links': {'first': '/api/content-sources/v1/tasks/?limit=100&offset=0',
           'last': '/api/content-sources/v1/tasks/?limit=100&offset=0'},
 'meta': {'count': 2, 'limit': 100, 'offset': 0}}
```

**Comment by swadeley** [comment]
hi
we can also GET the individual task info

```
In [1]: app.content_sources.rest_client.tasks_api.get_task('b8b72afd-0859-491b-9182-4de101fe0dab')
2023-06-23 15:42:39.996 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=BP84Ui30cLMXMCy1SmxmyzLxQzQD39tf, params=[]
Out[1]: 
{'created_at': '2023-06-23T14:05:44Z',
 'ended_at': '2023-06-23T14:05:44Z',
 'error': '',
 'org_id': '3340851',
 'status': 'completed',
 'uuid': 'b8b72afd-0859-491b-9182-4de101fe0dab'}

In [2]: app.content_sources.rest_client.tasks_api.get_task('e6fb679a-7eac-46b7-b2e6-78a077005c5e')
2023-06-23 15:43:03.491 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=JapXcX2MQZ6YMKo8Tz6cwmJAfuAKs4AA, params=[]
Out[2]: 
{'created_at': '2023-06-23T14:05:44Z',
 'ended_at': '2023-06-23T14:06:19Z',
 'error': '',
 'org_id': '3340851',
 'status': 'completed',
 'uuid': 'e6fb679a-7eac-46b7-b2e6-78a077005c5e'}
```

**Comment by swadeley** [comment]
Hi

Using the UUID of the repo, we can get info about snapshots of that repo:

```
In [5]: app.content_sources.rest_client.repositories_api.list_snapshots('60c3b07f-9f53-4c62-ac43-80c639edb30a')
2023-06-23 15:48:34.484 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=Lfv9BE5yCF9Z1CSNMk8t27eWNMD6mfgY, params=[]
Out[5]: 
{'data': [{'content_counts': {'rpm.package': 1},
           'created_at': '2023-06-23T14:06:19.383977Z',
           'distribution_path': '60c3b07f-9f53-4c62-ac43-80c639edb30a/2a3fea66-734f-4172-a8e2-b219c8a78990'}],
 'links': {'first': '/api/content-sources/v1/repositories/60c3b07f-9f53-4c62-ac43-80c639edb30a/snapshots/?limit=100&offset=0',
           'last': '/api/content-sources/v1/repositories/60c3b07f-9f53-4c62-ac43-80c639edb30a/snapshots/?limit=100&offset=0'},
 'meta': {'count': 1, 'limit': 100, 'offset': 0}}
```

**Pull request merged**